src/rpc/mining.cpp

```diff
@@ -123,10 +125,19 @@ UniValue generateBlocks(boost::shared_ptr<CReserveScript> coinbaseScript, int nG
           // LOCK(cs_main);
             IncrementExtraNonce(pblock, nExtraNonce);
         }
-        while (nMaxTries > 0 && pblock->nNonce < nInnerLoopCount && !CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus())) {
+
+        const uint32_t req_nBits = generateWeak > 0 ? WeakBlockProofOfWork(pblock->nBits) :pblock->nBits;
+
+        do {
             ++pblock->nNonce;
-            --nMaxTries;
-        }
+            while (nMaxTries > 0 && pblock->nNonce < nInnerLoopCount && !CheckProofOfWork(pblock->GetHash(), req_nBits, Params().GetConsensus())) {
+                ++pblock->nNonce;
+                --nMaxTries;
+            }
+            if (nMaxTries == 0 || pblock->nNonce >= nInnerLoopCount) break;
+            LogPrint("weakblocks", "Generated maybe strong block at difficulty %s, strong %s, weak %s, hash %s, nonce: %d.\n", arith_uint256().SetCompact(req_nBits).GetHex(), arith_uint256().SetCompact(pblock->nBits).GetHex(), arith_uint256().SetCompact(WeakBlockProofOfWork(pblock->nBits)).GetHex(), pblock->GetHash().GetHex(), pblock->nNonce);
+        } while (generateWeak == 2 && CheckProofOfWork(pblock->GetHash(), pblock->nBits, Params().GetConsensus())); // in mode 2, prevent strong blocks
```

Comment:
minor: is it worth creating an enum of modes so that we don't have that magic `2` floating around?

Reply:
Yes, makes sense. Will do.